core/distributed/dense_communicator.cpp

```diff
@@ -0,0 +1,166 @@
+// SPDX-FileCopyrightText: 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "ginkgo/core/distributed/dense_communicator.hpp"
+
+namespace gko {
+namespace experimental {
+namespace mpi {
+
+
+size_type get_comm_size_safe(const communicator& comm)
+{
+    if (comm.get() == MPI_COMM_NULL) {
+        return 0;
+    }
+    return comm.size();
+}
+
+
+DenseCommunicator::DenseCommunicator(communicator base)
+    : CollectiveCommunicator(base),
+      comm_(base),
+      recv_sizes_(get_comm_size_safe(comm_)),
+      recv_offsets_(get_comm_size_safe(comm_) + 1),
+      send_sizes_(get_comm_size_safe(comm_)),
+      send_offsets_(get_comm_size_safe(comm_) + 1)
+{}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+DenseCommunicator::DenseCommunicator(
+    communicator base,
+    const distributed::index_map<LocalIndexType, GlobalIndexType>& imap)
+    : DenseCommunicator(base)
+{
+    auto exec = imap.get_executor();
+    if (!exec) {
+        return;
+    }
+    auto host_exec = exec->get_master();
+
+    auto recv_target_ids_arr =
+        make_temporary_clone(host_exec, &imap.get_remote_target_ids());
+    auto remote_idx_offsets_arr = make_temporary_clone(
+        host_exec, &imap.get_remote_global_idxs().get_offsets());
+    for (size_type seg_id = 0;
+         seg_id < imap.get_remote_global_idxs().get_segment_count(); ++seg_id) {
+        recv_sizes_[recv_target_ids_arr->get_const_data()[seg_id]] =
+            remote_idx_offsets_arr->get_const_data()[seg_id + 1] -
+            remote_idx_offsets_arr->get_const_data()[seg_id];
+    }
+
+    comm_.all_to_all(host_exec, recv_sizes_.data(), 1, send_sizes_.data(), 1);
+
+    std::partial_sum(send_sizes_.begin(), send_sizes_.end(),
+                     send_offsets_.begin() + 1);
+    std::partial_sum(recv_sizes_.begin(), recv_sizes_.end(),
+                     recv_offsets_.begin() + 1);
```

Comment:
miss setting the first element of `send_offsets_` and `recv_offsets_`

Reply:
The first elements are already initialized to zero, as that is what the std::vector automatically does.